sql/memory.go

```diff
@@ -0,0 +1,189 @@
+package sql
+
+import (
+	"os"
+	"runtime"
+	"strconv"
+	"sync"
+
+	errors "gopkg.in/src-d/go-errors.v1"
+)
+
+// Disposable objects can erase all their content when they're no longer in use.
+// They should not be used again after they've been disposed.
+type Disposable interface {
+	// Dispose the contents.
+	Dispose()
+}
+
+// Freeable objects can free their memory.
+type Freeable interface {
+	// Free the memory.
+	Free()
+}
+
+// KeyValueCache is a cache of key value pairs.
+type KeyValueCache interface {
+	// Put a new value in the cache. If there is no more memory and the cache is
+	// not Freeable it will try to free some memory from other freeable caches.
+	// If there's still no more memory available, it will fail and erase all the
+	// contents of the cache.
+	// If it's Freeable, it will be freed, then the new value will be inserted.
+	Put(uint64, interface{}) error
+	// Get the value with the given key.
+	Get(uint64) (interface{}, error)
+}
+
+// RowsCache is a cache of rows.
+type RowsCache interface {
+	// Add a new row to the cache. If there is no memory available, it will try to
+	// free some memory. If after that there is still no memory available, it
+	// will return an error and erase all the content of the cache.
+	Add(Row) error
+	// Get all rows.
+	Get() []Row
+}
+
+// ErrNoMemoryAvailable is returned when there is no more available memory.
+var ErrNoMemoryAvailable = errors.NewKind("no memory available")
+
+const maxMemoryKey = "MAX_MEMORY"
+
+const (
+	b   = 1
+	kib = 1024 * b
+	mib = 1024 * kib
+)
+
+var maxMemory = func() uint64 {
+	v, _ := strconv.ParseUint(os.Getenv(maxMemoryKey), 10, 64)
```

Comment:
If it's not convertible, the function will return 0, which means not handled. On the other hand, it may actually be better to panic here if the value is misformatted, so the user does not think they set a max memory value and then be surprised by the fact that there is no limit.